[Data Cleaning] remove DATA_CLEANING_CASES toggle and release feature

biyeun opened the pull request:
## Product Description
Release Bulk Edit Case Data tool to Advanced plans and above.


## Safety Assurance

### Safety story
feature has been QA'd and thoroughly tested on staging.

### Automated test coverage
most important functionality is tested

### QA Plan
already completed
https://dimagi.atlassian.net/browse/QA-7508

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

biyeun committed:
remove DATA_CLEANING_CASES toggle and release feature

biyeun committed:
Merge branch 'master' into bmb/dc/release-cases

millerdev reviewed (commented):
Looks good, but test failures look related to the change, so I'm guessing this is not quite finished yet.

biyeun committed:
remove test, as check is no longer needed
- the flag is removed

biyeun committed:
Merge branch 'master' into bmb/dc/release-cases

biyeun committed:
move clear_caches_data_cleaning
- to the refresh_data_dictionary_from_app task that's called from the application save. the same trigger fires this task, just asynchronously. has the added event of not requiring a bunch of old tests to require the case_search_es index

millerdev commented on corehq/apps/app_manager/models.py:
```diff
@@ -4557,7 +4556,6 @@ def save(self, response_json=None, increment_version=None, **params):
         get_apps_in_domain.clear(self.domain, True)
         get_apps_in_domain.clear(self.domain, False)
         get_mobile_ucr_count.clear(self.domain)
-        clear_caches_case_data_cleaning(self.domain)
```
Moving this to an async task means the cache will not be cleared until after save, where this was previously done before save. In either case there is a race condition. It's not clear which is better, but just noting that such race conditions can cause hard-to-find bugs.

It's also notable that moving this one to after save means the race conditions are split since some caches are still being cleared before.

millerdev replied:
There is definitely a race condition. The question is whether it matters in practice. I have not done the research to know, and posed the question in hopes that you would.

mjriley commented on corehq/apps/data_dictionary/models.py:
```diff
@@ -151,7 +151,7 @@ def clear_caches(cls, domain, case_type):
         get_data_dict_props_by_case_type.clear(domain)
         get_gps_properties.clear(domain, case_type)
-        if toggles.DATA_CLEANING_CASES.enabled(domain):
+        if domain_has_privilege(domain, privileges.BULK_DATA_EDITING):
```
should clearing the caches be conditional on whether or not a domain has the privilege? It feels like this introduces an edge case where a domain has data cleaning, has an open session, then downgrades and loses privileges on data cleaning. They do something to trigger `clear_caches` (I'm guessing release a new app build), which, because they don't have the privilege anymore, does not clear the data cleaning cache. Then, when funding is restored, they upgrade and once again have data cleaning. At that point, they'd see the stale data from the cache, right?

biyeun replied:
it doesn't really matter, but i was following the pattern of the line below it, which also relies on a privledge

mjriley commented on corehq/apps/app_manager/tasks.py:
```diff
@@ -112,6 +113,7 @@ def refresh_data_dictionary_from_app(domain, app_id):
                         case_type_to_prop[action.case_type].update(action.case_properties)
     if case_type_to_prop:
         create_properties_for_case_types(domain, case_type_to_prop)
+    clear_caches_case_data_cleaning(domain)
```
Agreed, this feels strange, especially when a call like `expire_case_properties_caches` (and other clearing functions) are called in the main save method

millerdev reviewed (approved):
Leaving it up to you if the race conditions are a blocker.
